File: packages/core/src/node/state-adapter.js
const getStateManagerHandler = (stateManagerUrl) => async (action, payload) => {
  try {
    const response = await fetch(`${stateManagerUrl}/${action}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-trace-id': payload.traceId,
        // TODO: add internal auth token for security
      },
      // TODO: encrypt the payload for security
      body: JSON.stringify(payload),
    })

    if (!response.ok) {
      console.error('[internal state manager] failed posting state change, response ok status is false')
      throw Error('Failed posting state change, response ok status is false')
    }

    if (action === 'get') {
      const result = await response.json()
      return result.data
    }
  } catch (error) {
    console.error('[internal state manager] failed posting state change', error)
    throw Error('Failed posting state change')
  }
}

const createInternalStateManager = ({ stateManagerUrl }) => {
  const handler = getStateManagerHandler(stateManagerUrl)

  return {
    get: (traceId, key) => handler('get', { traceId, key }).then((result) => ({ data: result })),
    set: (traceId, key, value) => handler('set', { traceId, key, value }),
    delete: (traceId, key) => handler('delete', { traceId, key }),
  }
}

module.exports = {
  createInternalStateManager,
}

File: packages/core/src/ruby/state_adapter.rb
require 'net/http'
require 'json'
require 'uri'
require 'logger'

class StateManagerError < StandardError; end

LOGGER = Logger.new($stdout)

def get_state_manager_handler(state_manager_url)
  lambda do |action, payload|
    begin
      uri = URI("#{state_manager_url}/#{action}")
      request = Net::HTTP::Post.new(uri)
      request['Content-Type'] = 'application/json'
      request['x-trace-id'] = payload[:traceId]
      # TODO: Add internal auth token for security
      # TODO: Encrypt the payload for security
      request.body = payload.to_json

      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https') do |http|
        http.request(request)
      end

      unless response.is_a?(Net::HTTPSuccess)
        error_message = "Failed posting state change: #{response.code} - #{response.message}"
        LOGGER.error(error_message)
        raise StateManagerError, error_message
      end

      if action == 'get'
        result = JSON.parse(response.body, object_class: OpenStruct)
        return result.data
      end

      nil
    rescue StandardError => error
      LOGGER.error("[internal state manager] failed posting state change: #{error.message}")
      raise StateManagerError, "Failed posting state change: #{error.message}"
    end
  end
end

def create_internal_state_manager(state_manager_url:)
  handler = get_state_manager_handler(state_manager_url)

  {
    get: ->(trace_id, key) { 
      result = handler.call('get', { traceId: trace_id, key: key })
      { data: result }
    },
    set: ->(trace_id, key, value) { 
      handler.call('set', { traceId: trace_id, key: key, value: value })
    },
    delete: ->(trace_id, key) { 
      handler.call('delete', { traceId: trace_id, key: key })
    },
    clear: ->(trace_id) {
      handler.call('clear', { traceId: trace_id })
    }
  }
end
